### PR TITLE
Add support for SamrEnumerateDomainsInSamServer

### PR DESCRIPTION
**Pull request by bli-r7**

### CODE REVIEW

File: src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ *
+ */
+public class SamrEnumerateDomainsInSamServerRequest extends RequestCall<SamrEnumerateDomainsInSamServerResponse> {
+    public final static short OP_NUM = 6;
+    private final ServerHandle serverHandle;
+    private final int enumContext;
+    private final int maxLength;
+
+    public SamrEnumerateDomainsInSamServerRequest(ServerHandle handle, int enumContext, int maxLength) {
```
Comment (sangdo-lee) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java]: it it fine start with `{` ?

Comment (bli-r7) [reply src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java]: @sangdo-lee What do you mean?

Comment (sangdo-lee) [reply src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java]: oh sorry, never mind it. before I thought {  was started with new line..

Comment (bli-r7) [reply src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java]: Yea @tosmun-r7 had a big commit where we standardize our code format. We always keep the { on the same line from now on in smbj-rpc.

File: src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ **************************************************************************/
+package com.rapid7.client.dcerpc.mssamr.messages;
+
+import java.io.IOException;
+import com.rapid7.client.dcerpc.io.PacketOutput;
+import com.rapid7.client.dcerpc.messages.RequestCall;
+import com.rapid7.client.dcerpc.mssamr.objects.ServerHandle;
+
+/**
+ *
```
Comment (bli-r7) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java]: javadoc

File: src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/***************************************************************************
```
Comment (bli-r7) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/messages/SamrEnumerateDomainsInSamServerRequest.java]: fix header

File: src/main/java/com/rapid7/client/dcerpc/mssamr/objects/ServerHandle.java [hunk 4]
```diff
@@ -30,4 +30,8 @@
     public ServerHandle() {
         super();
     }
+
+    public ServerHandle(final String handle) {
```
Comment (bli-r7) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/objects/ServerHandle.java]: Revert this.

File: src/main/java/com/rapid7/client/dcerpc/mssamr/SecurityAccountManagerService.java [hunk 5]
```diff
@@ -79,4 +98,24 @@ public void closeHandle(ContextHandle handle) throws IOException {
...
+        final int bufferSize = 0xffff;
+        List<DomainInfo> domains = new ArrayList<>();
+        for (int enumContext = 0;;) {
+            SamrEnumerateDomainsInSamServerRequest request = new SamrEnumerateDomainsInSamServerRequest(serverHandle,
+                enumContext, bufferSize);
+            final SamrEnumerateDomainsInSamServerResponse response = transport.call(request);
+            final int returnCode = response.getReturnValue();
+            enumContext = response.getResumeHandle();
+            if (ERROR_MORE_ENTRIES.is(returnCode)) {
+                domains.addAll(response.getDomainList());
```
Comment (tosmun-r7) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/SecurityAccountManagerService.java]: This is efficient :+1:

File: src/main/java/com/rapid7/client/dcerpc/mssamr/objects/SAMPR_ENUMERATION_BUFFER.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public void unmarshalPreamble(PacketInput in) throws IOException {
+    }
+
+    @Override
+    public void unmarshalEntity(PacketInput in) throws IOException {
+        // Reference ID of sam_array
+        in.readReferentID();
```
Comment (tosmun-r7) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/objects/SAMPR_ENUMERATION_BUFFER.java]: Allow user to decide if this is a pointer or not. So move it out of this class and use it in your `SamrEnumerateDomainsInSamServerResponse`. This makes the class more portable

File: src/main/java/com/rapid7/client/dcerpc/mssamr/objects/SAMPR_ENUMERATION_BUFFER.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        // Entries of sam_array
+        entriesRead = in.readInt();
+        // Reference ID of domain_list
+        in.readReferentID();
+        // Entries of domain_list
+        int entryCount = in.readInt();
```
Comment (tosmun-r7) [inline src/main/java/com/rapid7/client/dcerpc/mssamr/objects/SAMPR_ENUMERATION_BUFFER.java]: From this line onward should be moved to `unmarshalDeferrals`. When reading a pointer (`[size_is(EntriesRead)] PSAMPR_RID_ENUMERATION Buffer;`), the only thing to be read is the referent ID.
However, you should also chose whether or not to initialize the array based on whether or not it is a null pointer.

For example:
```
@Override
    public void unmarshalEntity(PacketInput in) throws IOException {
        // Entries of sam_array
        entriesRead = in.readInt();
        // Reference ID of domain_list
        if (in.readReferentID() != 0) {
            if (entriesRead > 0)
                array = new ArrayList<>(entriesRead);
        }

    }

    @Override
    public void unmarshalDeferrals(PacketInput in) throws IOException {
        // Entries of domain_list
        if (array != null) {
            // MaximumCount
            for (int i = 0; i < in.readInt(); i++) {
                T t = initEntity();
                array.add(t);
                t.unmarshalPreamble(in);
            }
            for (T t : array) {
                t.unmarshalEntity(in);
            }
            for (T t : array) {
                t.unmarshalDeferrals(in);
            }
        }
    }
```

Comment (tosmun-r7) [reply src/main/java/com/rapid7/client/dcerpc/mssamr/objects/SAMPR_ENUMERATION_BUFFER.java]: Alignment may be required:
```
final Alignment tAlignment = initEntity().getAlignment();
in.align(Alignment.FOUR.getOffByOneAlignment() > tAlignment.getOffByOneAlignment() ? Alignment.FOUR : tAlignment);
```
But it does work for me regardless

File: src/test/java/com/rapid7/client/dcerpc/mssamr/Test_SecurityAccountManagerService.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
+/***************************************************************************
```
Comment (bli-r7) [inline src/test/java/com/rapid7/client/dcerpc/mssamr/Test_SecurityAccountManagerService.java]: header

File: src/test/java/com/rapid7/client/dcerpc/mssamr/messages/Test_SamrEnumerateDomainsInSamServerResponse.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
+/***************************************************************************
```
Comment (bli-r7) [inline src/test/java/com/rapid7/client/dcerpc/mssamr/messages/Test_SamrEnumerateDomainsInSamServerResponse.java]: header

**Pull request merged**